zeus/metric.py

```diff
@@ -0,0 +1,477 @@
+        )
+
+        self.bucket_ranges["cpu"] = (
+            bucket_ranges.get("cpu")
+            if bucket_ranges and "cpu" in bucket_ranges
+            else default_cpu_buckets
+        )
+
+        self.bucket_ranges["dram"] = (
+            bucket_ranges.get("dram")
+            if bucket_ranges and "dram" in bucket_ranges
+            else default_dram_buckets
+        )
+        # If GPU availble, for each gpu_indices, create a Histogram metric with the label window, and index.
+        if energy_monitor.gpu_indices:
+            self.gpu_histograms = {}
+            for gpu_index in self.energy_monitor.gpu_indices:
+                self.gpu_histograms[gpu_index] = Histogram(
+                    f"energy_monitor_gpu_{gpu_index}_energy_joules",
+                    f"GPU {gpu_index} energy consumption",
+                    ["window", "index"],
+                    buckets=self.bucket_ranges.get("gpu", []),
+                    registry=self.registry,
+                )
+        else:
+            self.gpu_histogram = None
+        # If CPU available, for each cpu_indices, create a Histogram metric with the label window, and index.
+        if energy_monitor.cpu_indices:
+            self.cpu_histograms = {}
+            for cpu_index in self.energy_monitor.cpu_indices:
+                self.cpu_histograms[cpu_index] = Histogram(
+                    f"energy_monitor_cpu_{cpu_index}_energy_joules",
+                    f"CPU {cpu_index} energy consumption",
+                    ["window", "index"],
+                    buckets=self.bucket_ranges.get("cpu", []),
+                    registry=self.registry,
+                )
+            # Only when CPUs are available, we check if DRAM is available using supportsGetDramEnergyConsumption in CPU class
+            # If DRAM available, we create histogram for each DRAM indices for each CPU indices
+            if CPU.supportsGetDramEnergyConsumption:
+                self.dram_histograms = {}
+                for dram_index in self.energy_monitor.cpu_indices:
+                    self.dram_histograms[dram_index] = Histogram(
+                        f"energy_monitor_dram_{dram_index}_energy_joules",
+                        f"DRAM {dram_index} energy consumption",
+                        ["window", "index"],
+                        buckets=self.bucket_ranges.get("dram", []),
+                        registry=self.registry,
+                    )
+            else:
+                self.dram_histogram = None
+        else:
+            self.cpu_histogram = None
+
+        self.max_gpu_bucket = max(self.bucket_ranges.get("gpu"))
+        self.max_cpu_bucket = max(self.bucket_ranges.get("cpu"))
+        self.max_dram_bucket = max(self.bucket_ranges.get("dram"))
+
+    def begin_window(self, name: str) -> None:
+        """Begin a new energy monitoring window."""
+        self.energy_monitor.begin_window(f"__EnergyHistogram_{name}")
+
+    def end_window(self, name: str) -> None:
```

Comment:
In all of the metrics, please mention somewhere that the name of the window **must** match between calls to `begin_window` and `end_window`.